### PR TITLE
Lex, Shine, and Parse

### PR DESCRIPTION
**Pull request by CodaFi**
Adds Layout and Parse to the Crust.

### CODE REVIEW

File: Sources/Crust/Parser.swift [hunk 1]
```diff
@@ -42,4 +50,543 @@ public class Parser {
...
+  func parseDeclList() -> DeclListSyntax {
+    var pieces = [DeclSyntax]()
+    while let piece = try? parseDecl() {
+      pieces.append(piece)
+    }
+    return DeclListSyntax(elements: pieces)
+  }
+
+  func parseDecl() throws -> DeclSyntax {
+    return try split({
```
Comment (harlanhaskins) [inline Sources/Crust/Parser.swift]: Why not switch on the token kind here? It’d be much easier to understand and way faster...

File: Sources/Crust/Parser.swift [hunk 2]
```diff
@@ -42,4 +50,543 @@ public class Parser {
...
+    let importTok = try consume(.importKeyword)
+    let ident = try parseQualifiedName()
+    return ImportDeclSyntax(importToken: importTok, importIdentifier: ident)
+  }
+}
+
+extension Parser {
+  func parseRecordDecl() throws -> RecordDeclSyntax {
+    let recordTok = try consume(.recordKeyword)
+    let recName = try parseQualifiedNamePiece().name
```
Comment (harlanhaskins) [inline Sources/Crust/Parser.swift]: Won’t this break the fidelity contract? We throw away the parent trivia. We need to update the grammar if this is wrong.

File: Sources/Crust/Parser.swift [hunk 3]
```diff
@@ -42,4 +50,543 @@ public class Parser {
...
+    }
+
+    guard !pieces.isEmpty else {
+      throw ParseError.failedParsingFunctionClause
+    }
+    return FunctionClauseListSyntax(elements: pieces)
+  }
+
+  func parseFunctionClause() throws -> FunctionClauseSyntax {
+    return try split({
```
Comment (harlanhaskins) [inline Sources/Crust/Parser.swift:331]: This pattern is also significantly more difficult to debug.

File: Sources/Crust/Parser.swift [hunk 4]
```diff
@@ -42,4 +50,543 @@ public class Parser {
...
+      return try self.parseWithRuleFunctionClause()
+    }, {
+      return try self.parseNormalFunctionClause()
+    })
+  }
+
+  func parseWithRuleFunctionClause() throws -> WithRuleFunctionClauseSyntax {
+    let functionName = try parseQualifiedNamePiece().name
```
Comment (harlanhaskins) [inline Sources/Crust/Parser.swift]: Same here. We can’t throw away any tokens.

File: Sources/Crust/Shine.swift [hunk 5]
```diff
@@ -0,0 +1,113 @@
+/// Parser.swift
+///
+/// Copyright 2017, The Silt Language Project.
+///
+/// This project is released under the MIT license, a copy of which is
+/// available in the repository.
+import Lithosphere
+
+private enum LayoutBlock {
+  case Implicit
```
Comment (harlanhaskins) [inline Sources/Crust/Shine.swift]: Lowercase, please

File: Sources/Crust/Shine.swift [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+/// Process a raw Silt token stream into a Stainless Silt token stream by
+/// inserting layout markers in the appropriate places.  This ensures that we
+/// have an explicitly-scoped input to the Parser before we even try to do a
+/// Scope Check.
+public func layout(_ ts : [TokenSyntax]) -> [TokenSyntax] {
+  var toks = ts
+  var stainlessToks = [TokenSyntax]()
+  var layoutBlockStack = [LayoutBlock]()
+  while !toks.isEmpty && toks[0].tokenKind != .eof {
```
Comment (harlanhaskins) [inline Sources/Crust/Shine.swift:29]: `while let tok = toks.removeFirst(), tok.tokenKind != .eof`?

File: Sources/Crust/Shine.swift [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+    // Finally, check to see if we've got a token on a new line.  If so, we
+    // need to insert a semicolon in the token stream if we can't find one.
+    //
+    // FIXME: This seems wrong in general.
+    if
+      let peekTok = toks.first,
+      let tokRange = tok.sourceRange,
+      let peekRange = peekTok.sourceRange,
+      tokRange.start.line != peekRange.start.line,
```
Comment (harlanhaskins) [inline Sources/Crust/Shine.swift]: Check the Trivia instead?

File: Sources/Lithosphere/SyntaxCollection.swift [hunk 8]
```diff
@@ -16,6 +16,19 @@ import Foundation
...
 /// versions of the collection with different children.
 public class SyntaxCollection<SyntaxElement: Syntax>: Syntax {
+  internal override init(root: SyntaxData, data: SyntaxData) {
+    super.init(root: root, data: data)
+  }
+
+  internal init(kind: SyntaxKind, elements: [SyntaxElement]) {
+    var list = [RawSyntax]()
+    for elt in elements {
+      list.append(elt.raw)
```
Comment (harlanhaskins) [inline Sources/Lithosphere/SyntaxCollection.swift]: `elements.map { $0.raw }`?

File: Sources/SyntaxGen/Generator.swift [hunk 9]
```diff
@@ -12,6 +12,13 @@ extension FileHandle: TextOutputStream {
     }
 }
 
+extension String {
+  func camelCased() -> String {
```
Comment (harlanhaskins) [inline Sources/SyntaxGen/Generator.swift:16]: We have `.lowercaseFirstWord` for this

**Review by harlanhaskins — CHANGES_REQUESTED** [review]

File: Sources/SyntaxGen/Generator.swift (hunk 9 above)
Comment (CodaFi) [reply Sources/SyntaxGen/Generator.swift:16]: Not defined anywhere I can find.

File: Sources/Crust/Shine.swift (hunk 6 above)
Comment (CodaFi) [reply Sources/Crust/Shine.swift:29]: `removeFirst()` traps, it doesn't return an optional

File: Sources/Crust/Parser.swift (hunk 2 above)
Comment (CodaFi) [reply Sources/Crust/Parser.swift]: Just lazy.  I'll fix it

File: Sources/Crust/Shine.swift (hunk 7 above)
Comment (CodaFi) [reply Sources/Crust/Shine.swift]: I would have to search the trivia on each token for a newline, right?

File: Sources/Crust/Parser.swift (hunk 3 above)
Comment (CodaFi) [reply Sources/Crust/Parser.swift:331]: We can't know whether or not we have a `with` clause until we've parsed through the entire LHS.  I can collapse `WithRuleFunctionClause` and `NormalFunctionClause` and just make the with clause optional.

Comment (CodaFi) [reply Sources/Crust/Parser.swift:331]: Never mind, can't do that.  We don't define `SyntaxKind`s for abstract derived classes of `Syntax` and so you can't have an `Expr?` node

File: Sources/SyntaxGen/Generator.swift (hunk 9 above)
Comment (harlanhaskins) [reply Sources/SyntaxGen/Generator.swift:16]: Ah, sorry, it’s lowercaseFirstLetter in String+Helpers

File: Sources/Crust/Parser.swift [hunk 10]
```diff
@@ -42,4 +50,563 @@ public class Parser {
...
+      return DataDeclSyntax(
+        dataToken: dataTok,
+        dataIdentifier: dataId.name,
+        typedParameterList: paramList,
+        typeIndices: indices,
+        whereToken: whereTok,
+        constructorList: constrList
+      )
+    } catch let e {
+      throw e
```
Comment (harlanhaskins) [inline Sources/Crust/Parser.swift]: Was this for debugging?

File: Sources/Crust/Parser.swift (hunk 3 above)
Comment (harlanhaskins) [reply Sources/Crust/Parser.swift:331]: This is fine, then. 👍

File: Sources/Crust/Shine.swift (hunk 6 above)
Comment (harlanhaskins) [reply Sources/Crust/Shine.swift:29]: Ah, garbage. Cool, this is fine then.

File: Sources/Drill/Options.swift [hunk 11]
```diff
@@ -9,12 +9,16 @@ import Foundation
 
 /// The mode the compiler will be executing in.
 public enum Mode: String {
-    /// The compiler will describe all the tokens in the source file, along with
-    /// the leading and trailing trivia.
-    case describeTokens = "describe-tokens"
```
Comment (harlanhaskins) [inline Sources/Drill/Options.swift:14]: Thanks for fixing this 😅 we should run swiftlint on CI

**Review by harlanhaskins — APPROVED** [review]
This is great!

File: Sources/Crust/Parser.swift (hunk 10 above)
Comment (CodaFi) [reply Sources/Crust/Parser.swift]: Yes...

**Comment by harlanhaskins** [comment]
Oh, this still needs diagnostics too.

**Comment by CodaFi** [comment]
Diagnostics and recovery need to happen in a separate patch.  I want to commit ScopeCheck after this

**Comment by CodaFi** [comment]
⛵️

**Pull request merged**